test: Extend functional tests for addr relay

mzumsande opened the pull request:
This extends the functional test `p2p_addr_relay.py`.
It adds test coverage for address relay involving outbound peers, tests for both outgoing and incoming `GETADDR` requests and tests for `-blocksonly` mode.

The initial refactors and some of the new tests were taken from Amiti Uttarwar's PR #21528 - they are general test improvements not directly tied to the change proposed there.

amitiuttarwar committed:
[test] Refactor the addr relay test to prepare for new tests

Moves setting up the addr message into a repeatable function, and breaks up the
existing tests into separate functions for legibility.

amitiuttarwar committed:
[test] Extract sending an addr message into a helper

Also reduces mocktime to prevent idle disconnects
Co-Authored-By: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
[test] Add tests for addr relay in -blocksonly mode

Co-Authored-By: Martin Zumsande <mzumsande@gmail.com>

mzumsande committed:
[test] Add tests for getaddr behavior

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande committed:
[test] Add address relay tests involving outbound peers

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

jonatack reviewed (commented):
Concept ACK

amitiuttarwar commented on test/functional/p2p_addr_relay.py:
```diff
@@ -122,6 +123,36 @@ def relay_tests(self):
+        # of the outbound peer which is often sent before the GETADDR response.
+        assert_equal(inbound_peer.num_ipv4_received, 0)
+
+        self.log.info('Check that further addr messages sent from an outbound peer are relayed')
```
s/further/subsequent? later? 

my brain keeps thinking "further" relates to how far the message propagates through network nodes.

mzumsande replied:
changed the wording

amitiuttarwar reviewed (approved):
thank you for working on this! glad to see this increased test coverage for addr handling. 

overall looks good to me, some small questions / wording suggestions. 

the commit message for 604128088e says "[test] Extract setting up addr message into a helper", but it really extracts _sending_ addr message, with the previous commit extracting the set up. this was my fault, but I just noticed it now 😛 

RE commit 6eb3024f14 adding coverage for the ignore-first-message-from-outbound-peer: since this is the current behavior, it seems reasonable to add coverage (and thanks for adding the explanation), but just want to mention that I find this bitcoin core behavior super weird and think we should consider changing it. Either bitcoin core should be consistent with itself and propagate those initial self announcements, or it shouldn't self announce at all. From researching the addr behavior of alternative clients for #21528,  I saw that a lot of nodes just do one early self-advertisement, so its unfortunate if bitcoin core nodes are just black-holing that as a side effect of getaddr logic. This is all out of scope of this PR but I'm mentioning since the test coverage highlights the behavior.

tACK 11c79242d2, reviewed the code & ran tests on every commit. happy to re-ack if you edit based on review.

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -82,6 +122,96 @@ def run_test(self):
+        self.log.info('Check that we send a getaddr message to outbound-full-relay peers')
+        full_outbound_peer = self.nodes[0].add_outbound_p2p_connection(GetAddrStore(), p2p_idx=0, connection_type="outbound-full-relay")
+        full_outbound_peer.sync_with_ping()
+        assert(full_outbound_peer.getaddr_received)
```
fixed

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -82,6 +122,96 @@ def run_test(self):
+        for i in range(1000):
+            first_octet = i >> 8
+            second_octet = i % 256
+            a = "{}.{}.1.1".format(first_octet, second_octet)
```
TIL, nice. Changed the two occurrences where the line is touched anyway.

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -105,6 +122,42 @@ def relay_tests(self):
 
+    def getaddr_tests(self):
+        self.log.info('Test getaddr behavior')
+        self.log.info('Check that we send a getaddr message to outbound-full-relay peers')
```
done

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -105,6 +122,42 @@ def relay_tests(self):
+        full_outbound_peer.sync_with_ping()
+        assert(full_outbound_peer.getaddr_received)
+
+        self.log.info('Check that we do not send a getaddr message to block-relay-only peers')
```
done

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -105,6 +122,42 @@ def relay_tests(self):
+            first_octet = i >> 8
+            second_octet = i % 256
+            a = "{}.{}.1.1".format(first_octet, second_octet)
+            self.nodes[0].addpeeraddress(a, 8333)
```
Actually, I copied this block from `p2p_getaddr_caching.py` :smile:

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -78,7 +78,7 @@ def setup_addr_msg(self, num):
     def send_addr_msg(self, source, msg, receivers):
         source.send_and_ping(msg)
         # pop m_next_addr_send timer
-        self.mocktime += 30 * 60
```
done, I agree.

mzumsande commented on test/functional/p2p_addr_relay.py:
```diff
@@ -122,6 +123,36 @@ def relay_tests(self):
+
+        self.log.info('Check address relay to outbound peers')
+        block_relay_peer = self.nodes[0].add_outbound_p2p_connection(GetAddrStore(), p2p_idx=1, connection_type="block-relay-only")
+        block_relay_peer.sync_with_ping()
```
No, the sync is done in `send_addr_msg` in the next line. Removed.

mzumsande commented:
@jonatack , @amitiuttarwar Thanks a lot for the reviews! I addressed all comments above.

> the commit message for [6041280](https://github.com/bitcoin/bitcoin/commit/604128088ebf7ea0ce795c9683e32ee40daac7ed) says "[test] Extract setting up addr message into a helper", but it really extracts _sending_ addr message, with the previous commit extracting the set up. this was my fault, but I just noticed it now 

I changed the commit message and also added a note why mocktime is changed in this commit.
 
> RE commit [6eb3024](https://github.com/bitcoin/bitcoin/commit/6eb3024f142b88ac6cb73ea1ee28b60b7f2590ba) adding coverage for the ignore-first-message-from-outbound-peer: since this is the current behavior, it seems reasonable to add coverage (and thanks for adding the explanation), but just want to mention that I find this bitcoin core behavior super weird and think we should consider changing it. Either bitcoin core should be consistent with itself and propagate those initial self announcements, or it shouldn't self announce at all. From researching the addr behavior of alternative clients for #21528, I saw that a lot of nodes just do one early self-advertisement, so its unfortunate if bitcoin core nodes are just black-holing that as a side effect of getaddr logic. This is all out of scope of this PR but I'm mentioning since the test coverage highlights the behavior.

 I agree, the `fGetAddr` flag preventing this was initially meant to prevent relaying GETADDR responses - at some time the typical order of messages must have changed, and now it typically affects self-announcements. I discovered this unintended behavior during #19794, which would have removed it. 

I [closed](https://github.com/bitcoin/bitcoin/pull/19794#issuecomment-692916364) that PR and didn't open a follow-up as originally planned because after reading additional BIP155 discussion (especially [this](https://github.com/bitcoin/bips/pull/907#issuecomment-692894687) post by sipa) I realized that even addrs that are part extraordinarily small GETADDR responses <=10 wouldn't get relayed due to  the 10min recency limit anyway (and if they are recent enough, why not relay them?). The [suggestion](https://github.com/bitcoin/bitcoin/pull/19794#issuecomment-691184578) of keeping the flag around and making sure that it would apply to GETADDR responses again didn't really convince me anymore after that. (as opposed to just removing the flag and relying on the content of the addr messages themselves instead of timing)

amitiuttarwar commented:
re-ACK a732ee353c, small diff based on code review 

> I agree, the fGetAddr flag preventing this was ... 

thanks for that explanation, I'll take a closer look!

brunoerg reviewed (approved):
tACK and reviewed the code (seems good).

tested it on MacOS 11.1, ran several times on every commit

maflcko reviewed (approved):
Concept ACK a732ee353c1922a1f9ca082775884d190893e0e9 🌊

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK a732ee353c1922a1f9ca082775884d190893e0e9 🌊
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUinXAv/fV9qz2YWTlsh45kR8eeKCPJBEjTUqUAad/RtPpxDGxd43H0EilfP8VzG
7SLZlK5N4uRuL9EaRqxwV7d+ByHiKf+ARdicmPd7gjNg+FDgyDs/hZk5g9XsGWUh
vOrEC3MDHdgWtgYMnHAE9bBLYM/8HnxBy23FaAk4GORmIo1DImKXmh/ABFhftZ62
gslGncJYt0m/QqCE85BspFtbwvR5nRrOlng1GkYKsKpOypH2xr4oZHttWQJSWDIE
oK8F6rXBWHX09Yh2uLpZ0reVXl44hKdUeoTzYxIh78zyPzMRFqQ+GWBNH7FvY8B3
YCtDv6/YuHqf+iExqri2Rgm5wVwQu9wBNf1APcoSzxVVIN2lBVMAlxoDlj7Ne3VS
bB4wZauKC4jFUmDGlFwrmX0iY7S4KouFAmM6F2JbW6QDYDIWWaLl4ti+uAtfaLCr
UddKbFtrhb7B5bZM/7o2TBrswflVNFwkTC99IFKAGB5z606NSFYy9dbuGMEFDfZm
JoLhOr8n
=PXgI
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `799645bbc5af6a827e9e28310a9332d8c563094883569c68faf6c2f08baa24c2  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108799645bbc5af6a827e9e28310a9332d8c563094883569c68faf6c2f08baa24c2f010e70493bbc6179623c698e64edb294c5408fff0107001d1d251456a2c1904b608ba7f7b9b08f1200931a796e9adcbea9b117d5d1f332744678dd51b6978badec51b10fd8df12fdf08f10460865a62f00873df76b76200601b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107f91898b4d87fd6f1b3464b81ff11d7b08f120b3486ff955e4a496357984f576ce067c3cedf700440528fe0b0e1be9e98b8b3508f1046086683ef008e90b7d782ab80f0c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d55e6c67470bead193da6ef90c51479008f1046086683df00868d0092e0a34f54e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010f62d4163383bbf6d2b75e9973783edfe08f1046086683ef0085efea72583cd042e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented on test/functional/p2p_addr_relay.py:
```diff
@@ -82,6 +122,95 @@ def run_test(self):
+        self.log.info('Check that we relay address messages')
+        addr_source = self.nodes[0].add_p2p_connection(P2PInterface())
+        msg = self.setup_addr_msg(2)
+        addr_source.send_and_ping(msg)
+        self.mocktime += 5 * 60
+        self.nodes[0].setmocktime(self.mocktime)
+        full_outbound_peer.sync_with_ping()
```
Any reason not to use the `send_addr_msg()` helper here?

mzumsande replied:
I missed this when combining test commits from different branches. 
@amitiuttarwar would you like to apply this refactor in  #21528 when you touch the test again?

jnewbery replied:
Done in #21785

jnewbery reviewed (commented):
utACK a732ee353c1922a1f9ca082775884d190893e0e9

Nice additions. Thank you!

MarcoFalke commented on test/functional/p2p_addr_relay.py:
```diff
@@ -82,6 +122,95 @@ def run_test(self):
+        self.log.info('Check that subsequent addr messages sent from an outbound peer are relayed')
+        msg2 = self.setup_addr_msg(2)
+        self.send_addr_msg(full_outbound_peer, msg2, [inbound_peer])
+        assert_equal(inbound_peer.num_ipv4_received, 2)
```
This fails:

* https://cirrus-ci.com/task/4959397244829696?logs=ci#L6318

mzumsande replied:
The expected addr message is sent directly after the assert has failed, so it looks like a timing issue:
` node0 2021-04-27T06:55:10.562357Z (mocktime: 2021-04-27T07:10:08Z) [net.cpp:2967] [PushMessage] sending addr (61 bytes) peer=9 ` (https://cirrus-ci.com/task/4959397244829696?logs=ci#L6320)
With `AVG_ADDRESS_BROADCAST_INTERVAL=30s`, I thought that a mocktime of 5 mins should be plenty to prevent Poisson timer fails - I think it's not that.
Possibly, there needs to be an additional wait after the mocktime to give the node more time to actually send out the addr message?

I'll look into this more closely later today!

MarcoFalke replied:
fixed in #21785
